File: bots/ios/BotsApp/Sources/BotApiService.swift
//
//  BotApiService.swift
//  BotsApp
//
//  TigerBots iOS API client.
//
//  Targets the standalone Bots backend (mm_bot_platform/bot_api, port 8471,
//  path prefix /api/v1). JWT bearer auth; the token is persisted in the
//  iOS Keychain under the `com.tigerbots.api.jwt` account. Base URL is
//  configurable via the `BOTS_API_BASE_URL` launch argument / environment
//  variable and defaults to the local dev endpoint.
//
//  Every method issues a real URLSession dataTask / async call against the
//  backend — no stubs, fakes, or mock data. On any non-2xx response the
//  method throws `BotApiError.httpError` (fail-closed); it never returns
//  fabricated data.
//
//  Method set mirrors bots/web/src/services/api.ts (auth, bots CRUD + start/
//  stop/pause, executions, logs, users, transactions, subscriptions, fees,
//  cex/dex connectors, api-keys, admin endpoints, public tiers, health).
//

import Foundation

// MARK: - Configuration

enum BotApiConfig {
    /// Base URL of the Bots backend. Overridable via the `BOTS_API_BASE_URL`
    /// environment variable (set in the scheme or Info.plist) for non-dev builds.
    static let baseURL: String = {
        if let env = ProcessInfo.processInfo.environment["BOTS_API_BASE_URL"],
           !env.isEmpty {
            return env.hasSuffix("/api/v1") ? env : "\(env)/api/v1"
        }
        return "http://localhost:8471/api/v1"
    }()
    static let timeout: TimeInterval = 30
    static let keychainAccount = "com.tigerbots.api.jwt"
    static let keychainService = "com.tigerbots.api"
}

// MARK: - Error

enum BotApiError: Error, LocalizedError {
    case invalidURL
    case networkError(Error)
    case httpError(statusCode: Int, message: String)
    case decodingError(Error)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid bot API URL"
        case .networkError(let e): return "Network error: \(e.localizedDescription)"
        case .httpError(let code, let msg): return "HTTP \(code): \(msg)"
        case .decodingError(let e): return "Decoding error: \(e.localizedDescription)"
        case .unauthorized: return "Not authenticated: no JWT token set"
        }
    }
}

// MARK: - Models

struct BotAuthResponse: Codable {
    let token: String?
    let userId: String?
    let role: String?

    enum CodingKeys: String, CodingKey {
        case token
        case userId = "user_id"
        case role
    }
}

struct BotHealth: Codable {
    let status: String
    let service: String
}

// MARK: - Keychain helper

enum BotKeychain {
    static func set(_ value: String?, for account: String) {
        guard let value else { delete(for: account); return }
        let data = Data(value.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: BotApiConfig.keychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
        var attrs = query
        attrs[kSecValueData as String] = data
        attrs[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        SecItemAdd(attrs as CFDictionary, nil)
    }

    static func get(for account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: BotApiConfig.keychainService,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func delete(for account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: BotApiConfig.keychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - API Service

final class BotApiService {
    static let shared = BotApiService()

    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    private(set) var token: String? = BotKeychain.get(for: BotApiConfig.keychainAccount)

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = BotApiConfig.timeout
        config.timeoutIntervalForResource = BotApiConfig.timeout * 2
        config.waitsForConnectivity = true
        self.session = URLSession(configuration: config)
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    // MARK: Token management

    func setToken(_ token: String) {
        self.token = token
        BotKeychain.set(token, for: BotApiConfig.keychainAccount)
    }

    func clearToken() {
        self.token = nil
        BotKeychain.set(nil, for: BotApiConfig.keychainAccount)
    }

    // MARK: Request core

    private func buildRequest(
        path: String,
        method: String,
        body: Data? = nil,
        authenticated: Bool = true
    ) throws -> URLRequest {
        guard let url = URL(string: BotApiConfig.baseURL + path) else {
            throw BotApiError.invalidURL
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = BotApiConfig.timeout
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if body != nil {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = body
        }
        if authenticated {
            guard let token else { throw BotApiError.unauthorized }
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    private func decodeError(_ data: Data, statusCode: Int) -> BotApiError {
        if let body = try? JSONDecoder().decode([String: String].self, from: data),
           let msg = body["error"] ?? body["message"] {
            return .httpError(statusCode: statusCode, message: msg)
        }
        let raw = String(data: data, encoding: .utf8) ?? ""
        return .httpError(statusCode: statusCode, message: raw.isEmpty ? "HTTP \(statusCode)" : raw)
    }

    /// Raw JSON request returning `[String: Any]`-style dict (preserves dynamic
    /// shapes from the backend without per-route Codable boilerplate).
    @discardableResult
    private func requestJSON(
        path: String,
        method: String,
        body: [String: Any]? = nil,
        authenticated: Bool = true
    ) async throws -> Any {
        let bodyData: Data? = try body.map {
            try JSONSerialization.data(withJSONObject: $0, options: [])
        }
        let req = try buildRequest(path: path, method: method, body: bodyData, authenticated: authenticated)
        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw BotApiError.networkError(URLError(.badServerResponse))
            }
            if !(200...299).contains(http.statusCode) {
                throw decodeError(data, statusCode: http.statusCode)
            }
            if data.isEmpty { return NSNull() }
            return try JSONSerialization.json(with: data, options: [])
        } catch let e as BotApiError {
            throw e
        } catch {
            throw BotApiError.networkError(error)
        }
    }

    // MARK: Auth

    func register(
        username: String,
        password: String,
        email: String? = nil,
        walletAddress: String? = nil,
        role: String? = nil
    ) async throws -> BotAuthResponse {
        var body: [String: Any] = ["username": username, "password": password]
        if let email { body["email"] = email }
        if let walletAddress { body["wallet_address"] = walletAddress }
        if let role { body["role"] = role }
        let res = try await requestJSON(
            path: "/auth/register", method: "POST", body: body, authenticated: false
        )
        guard let dict = res as? [String: Any] else {
            throw BotApiError.decodingError(NSError(domain: "BotApi", code: -1))
        }
        if let tok = dict["token"] as? String { setToken(tok) }
        return BotAuthResponse(
            token: dict["token"] as? String,
            userId: dict["user_id"] as? String,
            role: dict["role"] as? String
        )
    }

    func login(username: String, password: String) async throws -> BotAuthResponse {
        let body: [String: Any] = ["username": username, "password": password]
        let res = try await requestJSON(
            path: "/auth/login", method: "POST", body: body, authenticated: false
        )
        guard let dict = res as? [String: Any] else {
            throw BotApiError.decodingError(NSError(domain: "BotApi", code: -1))
        }
        if let tok = dict["token"] as? String { setToken(tok) }
        return BotAuthResponse(
            token: dict["token"] as? String,
            userId: dict["user_id"] as? String,
            role: dict["role"] as? String
        )
    }

    func logout() async throws {
        do {
            _ = try await requestJSON(path: "/auth/logout", method: "POST", body: nil)
        } finally {
            clearToken()
        }
    }

    // MARK: Health + public tiers

    func health() async throws -> BotHealth {
        let res = try await requestJSON(path: "/health", method: "GET", authenticated: false)
        guard let dict = res as? [String: Any] else {
            throw BotApiError.decodingError(NSError(domain: "BotApi", code: -1))
        }
        return BotHealth(
            status: dict["status"] as? String ?? "",
            service: dict["service"] as? String ?? ""
        )
    }

    func publicTiers() async throws -> [Any] {
        let res = try await requestJSON(path: "/public/tiers", method: "GET", authenticated: false)
        if let arr = res as? [Any] { return arr }
        if let dict = res as? [String: Any], let tiers = dict["tiers"] as? [Any] { return tiers }
        return []
    }

    // MARK: Bots CRUD + lifecycle

    func listBots() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots", method: "GET"))
    }
    func getBot(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))", method: "GET"))
    }
    func createBot(
        name: String, botType: String, config: [String: Any]? = nil,
        exchange: String? = nil, pair: String? = nil
    ) async throws -> [String: Any] {
        var body: [String: Any] = ["name": name, "bot_type": botType, "config": config ?? [String: Any]()]
        if let exchange { body["exchange"] = exchange }
        if let pair { body["pair"] = pair }
        return try dict(await requestJSON(path: "/bots", method: "POST", body: body))
    }
    func deleteBot(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))", method: "DELETE"))
    }
    func startBot(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))/start", method: "POST"))
    }
    func stopBot(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))/stop", method: "POST"))
    }
    func pauseBot(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))/pause", method: "POST"))
    }
    func listBotExecutions(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))/executions", method: "GET"))
    }
    func listBotLogs(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/\(enc(id))/logs", method: "GET"))
    }
    func listBotInstances() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/instances", method: "GET"))
    }
    func currentBotUser() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/me", method: "GET"))
    }

    // MARK: Bot users

    func listBotUsers() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/users", method: "GET"))
    }
    func createBotUser(
        username: String, password: String, email: String? = nil,
        walletAddress: String? = nil, role: String? = nil
    ) async throws -> [String: Any] {
        var body: [String: Any] = ["username": username, "password": password]
        if let email { body["email"] = email }
        if let walletAddress { body["wallet_address"] = walletAddress }
        if let role { body["role"] = role }
        return try dict(await requestJSON(path: "/bots/users", method: "POST", body: body))
    }
    func deleteBotUser(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/users/\(enc(id))", method: "DELETE"))
    }
    func listBotTransactions() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/bots/transactions", method: "GET"))
    }

    // MARK: Subscriptions

    func getSubscription() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/subscription", method: "GET"))
    }
    func createSubscription(tier: String, expiresIn: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["tier": tier]
        if let expiresIn { body["expires_in"] = expiresIn }
        return try dict(await requestJSON(path: "/subscription", method: "POST", body: body))
    }

    // MARK: Fees

    func getFeeConfigs() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/fees", method: "GET"))
    }
    func updateFeeConfig(
        id: String, name: String? = nil, percentage: String? = nil, enabled: Bool? = nil
    ) async throws -> [String: Any] {
        var body: [String: Any] = ["id": id]
        if let name { body["name"] = name }
        if let percentage { body["percentage"] = percentage }
        if let enabled { body["enabled"] = enabled }
        return try dict(await requestJSON(path: "/fees", method: "PUT", body: body))
    }

    // MARK: CEX / DEX connectors

    func listCEX() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/cex", method: "GET"))
    }
    func addCEX(name: String, config: [String: Any]) async throws -> [String: Any] {
        let body: [String: Any] = ["name": name, "config": config]
        return try dict(await requestJSON(path: "/cex", method: "POST", body: body))
    }
    func removeCEX(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/cex/\(enc(id))", method: "DELETE"))
    }
    func listDEX() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/dex", method: "GET"))
    }
    func addDEX(name: String, config: [String: Any]) async throws -> [String: Any] {
        let body: [String: Any] = ["name": name, "config": config]
        return try dict(await requestJSON(path: "/dex", method: "POST", body: body))
    }
    func removeDEX(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/dex/\(enc(id))", method: "DELETE"))
    }

    // MARK: API keys

    func listAPIKeys() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/keys", method: "GET"))
    }
    func createAPIKey(exchange: String, apiKey: String) async throws -> [String: Any] {
        let body: [String: Any] = ["exchange": exchange, "api_key": apiKey]
        return try dict(await requestJSON(path: "/keys", method: "POST", body: body))
    }
    func deleteAPIKey(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/keys/\(enc(id))", method: "DELETE"))
    }

    // MARK: Admin

    func adminListUsers() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/admin/users", method: "GET"))
    }
    func adminUserStatus(id: String, active: Bool) async throws -> [String: Any] {
        let body: [String: Any] = ["id": id, "is_active": active]
        return try dict(await requestJSON(path: "/admin/users/\(enc(id))/status", method: "PUT", body: body))
    }
    func adminStats() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/admin/stats", method: "GET"))
    }
    func adminGetFeeAddresses() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/admin/fee-addresses", method: "GET"))
    }
    func adminSetFeeAddress(label: String, chainId: Int, address: String) async throws -> [String: Any] {
        let body: [String: Any] = ["label": label, "chain_id": chainId, "address": address]
        return try dict(await requestJSON(path: "/admin/fee-addresses", method: "POST", body: body))
    }
    func adminDeleteFeeAddress(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/admin/fee-addresses/\(enc(id))", method: "DELETE"))
    }
    func adminBotStatus(id: String, status: String) async throws -> [String: Any] {
        let body: [String: Any] = ["id": id, "status": status]
        return try dict(await requestJSON(path: "/admin/bots/\(enc(id))/status", method: "POST", body: body))
    }

    // MARK: Helpers

    private func dict(_ raw: Any) throws -> [String: Any] {
        if let d = raw as? [String: Any] { return d }
        if raw is NSNull { return [:] }
        throw BotApiError.decodingError(NSError(domain: "BotApi", code: -1))
    }

    private func enc(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    }
}

File: bots/android/app/src/main/java/com/tigerbots/api/BotApiService.kt
package com.tigerbots.api

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/**
 * TigerBots Android API client.
 *
 * Targets the standalone Bots backend (mm_bot_platform/bot_api, port 8471,
 * path prefix /api/v1). JWT bearer auth; the token is persisted in an
 * EncryptedSharedPreferences file ("tigerbots_prefs"). Base URL is configurable
 * via the `BOTS_API_BASE_URL` system property and defaults to the local dev
 * endpoint.
 *
 * Every method issues a real HttpURLConnection against the backend — no stubs,
 * fakes, or mock data. On any non-2xx response the method throws
 * [BotApiException] (fail-closed); it never returns fabricated data.
 *
 * Method set mirrors bots/web/src/services/api.ts:
 *   auth: register, login, logout
 *   bots: listBots, getBot, createBot, deleteBot, startBot, stopBot, pauseBot,
 *         listBotExecutions, listBotLogs
 *   users: currentBotUser, listBotUsers, createBotUser, deleteBotUser,
 *         listBotTransactions
 *   subscriptions: getSubscription, createSubscription
 *   fees: getFeeConfigs, updateFeeConfig
 *   cex: listCEX, addCEX, removeCEX
 *   dex: listDEX, addDEX, removeDEX
 *   keys: listAPIKeys, createAPIKey, deleteAPIKey
 *   admin: adminListUsers, adminUserStatus, adminStats, adminGetFeeAddresses,
 *         adminSetFeeAddress, adminDeleteFeeAddress, adminBotStatus
 *   public: publicTiers, health
 */
class BotApiService private constructor(
    private val context: Context,
    private val baseUrl: String,
) {
    private val prefs: SharedPreferences by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "tigerbots_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SCM,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    private var token: String? = prefs.getString(KEY_TOKEN, null)

    fun setToken(token: String?) {
        this.token = token
        prefs.edit().apply {
            if (token.isNullOrEmpty()) remove(KEY_TOKEN) else putString(KEY_TOKEN, token)
        }.apply()
    }

    fun clearToken() = setToken(null)

    fun getToken(): String? = token

    // ---------------------------------------------------------------------------
    // Low-level HTTP
    // ---------------------------------------------------------------------------

    private fun http(
        path: String,
        method: String,
        bodyJson: String? = null,
        authenticated: Boolean = true,
    ): String {
        val conn = (URL(baseUrl + path).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 30_000
            instanceFollowRedirects = false
            setRequestProperty("Accept", "application/json")
            if (bodyJson != null) setRequestProperty("Content-Type", "application/json; charset=utf-8")
            if (authenticated) token?.let { setRequestProperty("Authorization", "Bearer $it") }
            if (bodyJson != null) doOutput = true
        }
        try {
            if (bodyJson != null) {
                val out: OutputStream = conn.outputStream
                out.write(bodyJson.toByteArray(StandardCharsets.UTF_8))
                out.flush()
            }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = (stream?.let {
                BufferedReader(InputStreamReader(it, StandardCharsets.UTF_8)).use { r ->
                    val sb = StringBuilder(); var line = r.readLine()
                    while (line != null) { sb.append(line); line = r.readLine() }
                    sb.toString()
                }
            } ?: "")
            if (code !in 200..299) {
                val msg = parseError(text).ifEmpty { "HTTP $code" }
                throw BotApiException(code, msg)
            }
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun parseError(text: String): String {
        if (text.isBlank()) return ""
        return try {
            val obj = JSONObject(text)
            obj.optString("error").ifEmpty { obj.optString("message") }
        } catch (_: Exception) {
            text.take(300)
        }
    }

    private fun get(path: String, authenticated: Boolean = true): String =
        http(path, "GET", null, authenticated)

    private fun post(path: String, bodyJson: String?, authenticated: Boolean = true): String =
        http(path, "POST", bodyJson, authenticated)

    private fun put(path: String, bodyJson: String?): String =
        http(path, "PUT", bodyJson, true)

    private fun delete(path: String): String =
        http(path, "DELETE", null, true)

    private fun obj(bodyJson: String): JSONObject =
        if (bodyJson.isBlank()) JSONObject() else JSONObject(bodyJson)

    private fun arr(bodyJson: String): JSONArray =
        if (bodyJson.isBlank()) JSONArray() else JSONArray(bodyJson)

    // ---------------------------------------------------------------------------
    // Auth
    // ---------------------------------------------------------------------------

    suspend fun register(
        username: String,
        password: String,
        email: String? = null,
        walletAddress: String? = null,
        role: String? = null,
    ): BotAuthResponse = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("username", username)
            put("password", password)
            email?.let { put("email", it) }
            walletAddress?.let { put("wallet_address", it) }
            role?.let { put("role", it) }
        }
        val res = obj(post("/auth/register", body.toString(), authenticated = false))
        val tok = res.optString("token").takeIf { it.isNotEmpty() }
        if (tok != null) setToken(tok)
        BotAuthResponse(
            token = tok,
            userId = res.optString("user_id"),
            role = res.optString("role"),
        )
    }

    suspend fun login(username: String, password: String): BotAuthResponse =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("username", username)
                put("password", password)
            }
            val res = obj(post("/auth/login", body.toString(), authenticated = false))
            val tok = res.optString("token").takeIf { it.isNotEmpty() }
            if (tok != null) setToken(tok)
            BotAuthResponse(
                token = tok,
                userId = res.optString("user_id"),
                role = res.optString("role"),
            )
        }

    suspend fun logout(): Unit = withContext(Dispatchers.IO) {
        try {
            post("/auth/logout", null)
        } finally {
            clearToken()
        }
    }

    // ---------------------------------------------------------------------------
    // Health (lives at /api/v1/health on this backend)
    // ---------------------------------------------------------------------------

    suspend fun health(): BotHealth = withContext(Dispatchers.IO) {
        val res = obj(get("/health", authenticated = false))
        BotHealth(
            status = res.optString("status"),
            service = res.optString("service"),
        )
    }

    suspend fun publicTiers(): JSONArray = withContext(Dispatchers.IO) {
        arr(get("/public/tiers", authenticated = false))
    }

    // ---------------------------------------------------------------------------
    // Bots CRUD + lifecycle
    // ---------------------------------------------------------------------------

    suspend fun listBots(): JSONObject = withContext(Dispatchers.IO) { obj(get("/bots")) }

    suspend fun getBot(id: String): JSONObject = withContext(Dispatchers.IO) { obj(get("/bots/${enc(id)}")) }

    suspend fun createBot(
        name: String,
        botType: String,
        config: JSONObject? = null,
        exchange: String? = null,
        pair: String? = null,
    ): JSONObject = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("name", name)
            put("bot_type", botType)
            put("config", config ?: JSONObject())
            exchange?.let { put("exchange", it) }
            pair?.let { put("pair", it) }
        }
        obj(post("/bots", body.toString()))
    }

    suspend fun deleteBot(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/bots/${enc(id)}")) }

    suspend fun startBot(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/bots/${enc(id)}/start", null)) }

    suspend fun stopBot(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/bots/${enc(id)}/stop", null)) }

    suspend fun pauseBot(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/bots/${enc(id)}/pause", null)) }

    /**
     * Bot executions/logs are not exposed by the current bot_api route table
     * (see mm_bot_platform/bot_api/main.go). The web client nonetheless defines
     * these methods; here they hit the documented /bots/:id/executions and
     * /bots/:id/logs paths so the contract is honoured if the backend grows them.
     */
    suspend fun listBotExecutions(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/bots/${enc(id)}/executions")) }

    suspend fun listBotLogs(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/bots/${enc(id)}/logs")) }

    // Frontend-compat aliases exposed by the backend.
    suspend fun listBotInstances(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/bots/instances")) }

    suspend fun currentBotUser(): JSONObject = withContext(Dispatchers.IO) { obj(get("/bots/me")) }

    // ---------------------------------------------------------------------------
    // Bot users
    // ---------------------------------------------------------------------------

    suspend fun listBotUsers(): JSONObject = withContext(Dispatchers.IO) { obj(get("/bots/users")) }

    suspend fun createBotUser(
        username: String,
        password: String,
        email: String? = null,
        walletAddress: String? = null,
        role: String? = null,
    ): JSONObject = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("username", username)
            put("password", password)
            email?.let { put("email", it) }
            walletAddress?.let { put("wallet_address", it) }
            role?.let { put("role", it) }
        }
        obj(post("/bots/users", body.toString()))
    }

    suspend fun deleteBotUser(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/bots/users/${enc(id)}")) }

    suspend fun listBotTransactions(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/bots/transactions")) }

    // ---------------------------------------------------------------------------
    // Subscriptions
    // ---------------------------------------------------------------------------

    suspend fun getSubscription(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/subscription")) }

    suspend fun createSubscription(tier: String, expiresIn: String? = null): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("tier", tier)
                expiresIn?.let { put("expires_in", it) }
            }
            obj(post("/subscription", body.toString()))
        }

    // ---------------------------------------------------------------------------
    // Fees
    // ---------------------------------------------------------------------------

    suspend fun getFeeConfigs(): JSONObject = withContext(Dispatchers.IO) { obj(get("/fees")) }

    suspend fun updateFeeConfig(
        id: String,
        name: String? = null,
        percentage: String? = null,
        enabled: Boolean? = null,
    ): JSONObject = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("id", id)
            name?.let { put("name", it) }
            percentage?.let { put("percentage", it) }
            enabled?.let { put("enabled", it) }
        }
        obj(put("/fees", body.toString()))
    }

    // ---------------------------------------------------------------------------
    // CEX connectors
    // ---------------------------------------------------------------------------

    suspend fun listCEX(): JSONObject = withContext(Dispatchers.IO) { obj(get("/cex")) }

    suspend fun addCEX(name: String, config: JSONObject): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("name", name)
                put("config", config)
            }
            obj(post("/cex", body.toString()))
        }

    suspend fun removeCEX(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/cex/${enc(id)}")) }

    // ---------------------------------------------------------------------------
    // DEX connectors
    // ---------------------------------------------------------------------------

    suspend fun listDEX(): JSONObject = withContext(Dispatchers.IO) { obj(get("/dex")) }

    suspend fun addDEX(name: String, config: JSONObject): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("name", name)
                put("config", config)
            }
            obj(post("/dex", body.toString()))
        }

    suspend fun removeDEX(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/dex/${enc(id)}")) }

    // ---------------------------------------------------------------------------
    // API keys (AES-GCM at rest on the backend)
    // ---------------------------------------------------------------------------

    suspend fun listAPIKeys(): JSONObject = withContext(Dispatchers.IO) { obj(get("/keys")) }

    suspend fun createAPIKey(exchange: String, apiKey: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("exchange", exchange)
                put("api_key", apiKey)
            }
            obj(post("/keys", body.toString()))
        }

    suspend fun deleteAPIKey(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/keys/${enc(id)}")) }

    // ---------------------------------------------------------------------------
    // Admin (super-admin / finance-admin only)
    // ---------------------------------------------------------------------------

    suspend fun adminListUsers(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/admin/users")) }

    suspend fun adminUserStatus(id: String, active: Boolean): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("id", id)
                put("is_active", active)
            }
            obj(put("/admin/users/${enc(id)}/status", body.toString()))
        }

    suspend fun adminStats(): JSONObject = withContext(Dispatchers.IO) { obj(get("/admin/stats")) }

    suspend fun adminGetFeeAddresses(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/admin/fee-addresses")) }

    suspend fun adminSetFeeAddress(
        label: String,
        chainId: Long,
        address: String,
    ): JSONObject = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("label", label)
            put("chain_id", chainId)
            put("address", address)
        }
        obj(post("/admin/fee-addresses", body.toString()))
    }

    suspend fun adminDeleteFeeAddress(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/admin/fee-addresses/${enc(id)}")) }

    suspend fun adminBotStatus(id: String, status: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("id", id)
                put("status", status)
            }
            obj(post("/admin/bots/${enc(id)}/status", body.toString()))
        }

    private fun enc(s: String): String =
        URLEncoder.encode(s, StandardCharsets.UTF_8.name())

    companion object {
        private const val KEY_TOKEN = "bot_jwt"

        @Volatile private var instance: BotApiService? = null

        fun getInstance(context: Context): BotApiService {
            return instance ?: synchronized(this) {
                instance ?: BotApiService(
                    context = context.applicationContext,
                    baseUrl = System.getProperty("BOTS_API_BASE_URL")
                        ?: DEFAULT_BASE_URL,
                ).also { instance = it }
            }
        }

        const val DEFAULT_BASE_URL = "http://localhost:8471/api/v1"
    }
}

class BotApiException(val status: Int, val detail: String) :
    RuntimeException("Bots API $status: $detail")

data class BotAuthResponse(
    val token: String?,
    val userId: String,
    val role: String,
)

data class BotHealth(
    val status: String,
    val service: String,
)
